### DISCUSSION TITLE
Transformations applied to audio datasets

### DISCUSSION THREAD

**Discussion by msbouanane** [neuromorphs/tonic#168] [Q&A] [answered]
Hello, 
thank you very much for this wonderful package.  
I'm trying to train my SNN with the N-TIDIGITS audio dataset but I'm not sure about the transformations I should apply to it before passing it through the network. 
Anyone has any suggestions?

**Comment by biphasic** [comment]
Hey,
glad you find it useful. @sheiksadique is currently working on this. Please have a look at this branch https://github.com/synsense/tonic/tree/audio_transforms and consider merging the latest develop branch into that branch before using it. It's based on PyTorch audio and SciPy. We're planning to include that in the next release but it'll still take some time. What transforms are you looking for in particular?

**Reply by msbouanane** [reply]
I'm not sure about what transforms I should use for N-TIDIGITS. I trained my SNN on the NMNIST before and I used the `.ToFrame` transform. But for N-TIDIGITS, I don't know what kind of event transformations I need. I tried to train the network without applying any transforms to the dataset but I run into errors. I should note that I'm using snnTorch.

**Reply by biphasic** [reply]
I see. In principle, the transform you need is a similar one as ToFrame, which adds up all the events in a given time/event count window. Currently, ToFrame supports 3D (time, x, y) data and for audio data the "y" channel is missing, that's why it's complaining. This would need to be changed  [here](https://github.com/neuromorphs/tonic/blob/develop/tonic/functional/to_frame.py#L11). I will open a new issue and tag you. Do you think you'd like to look into this?

**Reply by msbouanane** [reply]
Yes of course

**Reply by biphasic** [reply]
thanks a lot. I opened an issue [here](https://github.com/neuromorphs/tonic/issues/169)

**Comment by biphasic** [comment, marked as answer]
I fixed this issue. You should now be able to use the ToFrame transform and choose a slicing method such as time_window, event_count or else to rasterize your data https://github.com/neuromorphs/tonic/commit/afaf3e95b485ed4124989f1bf69689a009df3ce6

**Comment by biphasic** [comment]
also here is an example file of how I used SHD in the past. This is a PyTorch lightning datamodule but you can see the transforms one might want to apply.

```python
import torch, math
import pytorch_lightning as pl
from tonic import datasets, transforms
from torch.utils.data import DataLoader
import numpy as np
import tonic


class ToRaster():
    def __init__(self, encoding_dim):
        self.encoding_dim = encoding_dim

    def __call__(self, events):
        # tensor has dimensions (time_steps, encoding_dim)
        tensor = np.zeros((events["t"].max()+1, self.encoding_dim), dtype=int)
        np.add.at(tensor, (events["t"], events["x"]), 1)
        return tensor[:250,:]


class HSD(pl.LightningDataModule):
    def __init__(self, batch_size, encoding_dim, dt=4000, num_workers=6, download_dir='./data'):
        super().__init__()
        self.batch_size = batch_size
        self.num_workers = num_workers
        self.download_dir = download_dir
        
        # SHD is recorded using 700 channels, which is a lot so we are going to 
        # downsample that to our liking. Similarly the timestamps in us resolution
        # need to be downsampled.
        self.transform = transforms.Compose([
            transforms.Downsample(time_factor=1/dt, spatial_factor=encoding_dim/700),
            ToRaster(encoding_dim),
        ])
  
    def prepare_data(self):
        datasets.SHD(self.download_dir, train=True)
        datasets.SHD(self.download_dir, train=False)
  
    def setup(self, stage=None):
        self.train_data = datasets.SHD(self.download_dir, train=True, transform=self.transform)
        self.test_data = datasets.SHD(self.download_dir, train=False, transform=self.transform)

    def train_dataloader(self):
        return DataLoader(self.train_data, num_workers=self.num_workers, batch_size=self.batch_size, 
                          collate_fn=tonic.collation.PadTensors(batch_first=True), drop_last=True, shuffle=True)
  
    def val_dataloader(self):
        return DataLoader(self.test_data, num_workers=self.num_workers, batch_size=self.batch_size, 
                          collate_fn=tonic.collation.PadTensors(batch_first=True), drop_last=True)

    def test_dataloader(self):
        return DataLoader(self.test_data, num_workers=self.num_workers, batch_size=self.batch_size, 
                          collate_fn=tonic.collation.PadTensors(batch_first=True), drop_last=True)
```

**Comment by msbouanane** [comment]
Thank you very much for your help! 
However I still have another inquiry regarding ToFrame Transform and audio datasets when used with snnTorch. When I applied ToFrame to the NMNIST dataset, it returns tensors with shape [ T , dim ] (e.g. `(298, 2, 34, 34)` with "dim" being the sensor_size tuple). Then when I pass it to the DataLoader it returns tensors with shape [ T , B , dim] (e.g. `torch.Size([310, 128, 2, 34, 34])`). This makes it easier to pass the dataloader to the `snntorch.backprop` module to train the network. 
However, when I apply ToFrame to an audio dataset like SHD, it doesn't return the TIME dimension in the event tensors but only the dimensions of the sensor_size.  
Without the time dimension, it is not possible to train the network using snnTorch's `backprop.BPTT`. 
Here are 2 screenshots to see how the transform responds differently to NMNIST and SHD or NTIDIGITS: 
![NMNIST](https://user-images.githubusercontent.com/95038580/146942329-3507bd98-38d1-42d4-b811-fbcc0445776b.png)
![SHD](https://user-images.githubusercontent.com/95038580/146942337-4d05cf75-87f8-4a4f-a9b8-6e87d4a9ae91.png)

**Comment by biphasic** [comment]
see answere here: https://github.com/neuromorphs/tonic/discussions/174#discussioncomment-1888620
